
# TODO [by 30 July] 
#  - [ ] Update `build-iht-based-haplotype-map-and-phase-variants.sh` to:
    - Ensure `gtg-ped-map`, etc. are in the PATH
    - Run `bgzip`, etc.
# update notebook to reflect new variable and function names and signatures 
# ask AI to generate a snakemake file from the manual workflow
# share repo with Tom N so that he can use it to phase Tom S's DNMs at TR loci to founder labels, and thereby compare with Tom S's current phasing methodology. 

# TODO [Aug: lab symposium] 
# [positive control] does my tool phase methylation levels to the correct parental haplotype at these imprinted loci: https://gemini.google.com/app/db2d1c5aa2a8ed10 ? 
# present phased methylation levels for a pedigree, 
# create a python library called "tapestry" for analyzing the phased meth data across generations, such as: given a haplotype in a child, has its methylation changed relative to the same haplotype in the parent; etc.
# and solicit feedback about other interesting questions to ask
