{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c26bec1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO Mon 11/24\n",
    "\n",
    "# 1. integrate the following into src/expand_to_all_cpgs.200081.ipynb \n",
    "#    point out the need to identify/flag CpG creation/destruction by referencing: \n",
    "#    slides: https://docs.google.com/presentation/d/11Pfax0wXh0E68C287lMaPoPvhq-OrFGxFOKE1gWOkDI/edit?slide=id.g39893c07c75_0_0#slide=id.g39893c07c75_0_0 \n",
    "#    slack thread: https://quinlangroup.slack.com/archives/C0803TM7X0X/p1762565840460019?thread_ts=1759348751.929209&cid=C0803TM7X0X \n",
    "\n",
    "# 3. load up variants (use, as template, get_iht_phasing from src/phase_meth_to_founder_haps.ipynb that generates a df of variants from \n",
    "#    https://github.com/quinlan-lab/tapestry/blob/5b5cdc7d3d48e4daa374dba17de4ac341af5fcf3/phase_meth_to_founder_haps.sh#L26), and filter to SNVs (for now)\n",
    "# 4. intersect (or join) DF_METH_FOUNDER_PHASED_ALL_CPGS and variant df to find those CpG sites that overlap SNVs (both het and hom alt), and label those CpG sites accordingly\n",
    "#    NOTE: dont replace corresponding meth values of 0.0 with None since \n",
    "#          (i) we use non-null values to determine whether phasing was possible and \n",
    "#          (ii) pb-cpg-tools reports non-zero values for model-based methylation at non-CpG sites! \n",
    "\n",
    "# 5. integrate new code in src/expand_to_all_cpgs.200081.ipynb into expand_to_all_cpgs.py, and rerun that final tapestry step (expand_to_all_cpgs.sh) for each sample\n",
    "# 6. update https://github.com/quinlan-lab/tapestry?tab=readme-ov-file#step-4-of-workflow-expanding-the-dataset-to-include-all-cpg-sites-in-the-reference-and-the-sample-genome-and-to-include-unphased-dna-methylation\n",
    "# 7. re-run analysis/imprinting.ipynb, but this time filtering out CpG sites that have been created or destroyed \n",
    "#    Nice to haves: \n",
    "#      i. can compute_methylation_all_samples_at_given_loci be combined with compute_delta_methylation_all_samples via refactoring? \n",
    "#      ii. label imprinted_meth_all_samples.bed with value of DELTA_METH_THRESHOLD"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv (3.11.11)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
