{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e73689ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bioframe as bf # https://bioframe.readthedocs.io/en/latest/index.html\n",
    "import polars as pl \n",
    "import sys\n",
    "from pathlib import Path\n",
    "import importlib\n",
    "\n",
    "REFERENCE_GENOME = \"hg38\"\n",
    "TILE_SIZE = 1000\n",
    "METH_READ_PHASED_DIR = Path('/scratch/ucgd/lustre-labs/quinlan/data-shared/dna-methylation/CEPH1463.GRCh38.hifi.founder-phased.all-cpgs')\n",
    "SAMPLE = '200081' \n",
    "BED_METH = f\"{METH_READ_PHASED_DIR}/{SAMPLE}.dna-methylation.founder-phased.all_cpgs.bed\"\n",
    "\n",
    "REPO_DIR = Path(\"/scratch/ucgd/lustre-labs/quinlan/u6018199/tapestry\")\n",
    "sys.path.append(str(REPO_DIR / \"src\" / \"util\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e66fd74f",
   "metadata": {},
   "source": [
    "## Tile the genome "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f35fc009",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tile\n",
    "importlib.reload(tile)\n",
    "from tile import get_tiles \n",
    "\n",
    "DF_TILES = get_tiles(REFERENCE_GENOME, TILE_SIZE)\n",
    "DF_TILES"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa0e8298",
   "metadata": {},
   "source": [
    "## Fetch phased and unphased DNA methylation at all CpG sites in the reference and sample genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1b8ed6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import read_data\n",
    "importlib.reload(read_data)\n",
    "from read_data import read_tapestry\n",
    "\n",
    "DF_METH = read_tapestry(BED_METH)\n",
    "DF_METH"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffe65337",
   "metadata": {},
   "source": [
    "## Compute difference in methylation between haplotypes in tiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c359eaa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import compute_methylation\n",
    "importlib.reload(compute_methylation)\n",
    "from compute_methylation import compute_methylation\n",
    "\n",
    "DF_TILES_WITH_METH = compute_methylation(DF_TILES, DF_METH)\n",
    "DF_TILES_WITH_METH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16b808ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO \n",
    "# 3. compute delta_meth for each tile \n",
    "# 4. full-join tiles over all samples, keeping meth_pat, meth_mat, delta_meth and hap_pat, hap_mat  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8628438",
   "metadata": {},
   "source": [
    "## Find imprinted loci and check for consistency across samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70293a84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5. filter tiles for those in which delta_meth is close to 1 in at least one sample \n",
    "# 6. is delta_meth the same for all samples? is meth_pat and meth_mat the same over all samples? compare with \"truth set\" of imprinted loci here: https://quinlangroup.slack.com/archives/C0803TM7X0X/p1758742312373849\n",
    "# 7. are there polymorphic loci, where delta-meth varies over samples? and if so, does meth_pat, meth_mat, delta_meth covary with founder haplotype of degree of relationship? \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv (3.11.11)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
