{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a9154eb5",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "409e229b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/usr/lib64/python311.zip',\n",
       " '/usr/lib64/python3.11',\n",
       " '/usr/lib64/python3.11/lib-dynload',\n",
       " '',\n",
       " '/scratch/ucgd/lustre-labs/quinlan/u6018199/tapestry/.venv/lib64/python3.11/site-packages',\n",
       " '/scratch/ucgd/lustre-labs/quinlan/u6018199/tapestry/.venv/lib64/python3.11/site-packages/cyvcf2-0.31.1-py3.11-linux-x86_64.egg',\n",
       " '/scratch/ucgd/lustre-labs/quinlan/u6018199/tapestry/.venv/lib/python3.11/site-packages',\n",
       " '/scratch/ucgd/lustre-labs/quinlan/u6018199/tapestry/.venv/lib/python3.11/site-packages/cyvcf2-0.31.1-py3.11-linux-x86_64.egg',\n",
       " '/scratch/ucgd/lustre-labs/quinlan/u6018199/tapestry/src/util']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "from pathlib import Path\n",
    "\n",
    "REFERENCE_GENOME = \"hg38\"\n",
    "\n",
    "REPO_DIR = Path(\"/scratch/ucgd/lustre-labs/quinlan/u6018199/tapestry\")\n",
    "sys.path.append(str(REPO_DIR / \"src\" / \"util\"))\n",
    "sys.path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d350b776",
   "metadata": {},
   "source": [
    "## Palladium Trio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "edb6372f",
   "metadata": {},
   "outputs": [],
   "source": [
    "FATHER_UID = \"NA12877\" \n",
    "MOTHER_UID = \"NA12878\"\n",
    "CHILD_UID = \"NA12881\" # 2211"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d57774d",
   "metadata": {},
   "source": [
    "## TR DNMs in child"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8fe8f91d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (91, 10)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>start</th><th>end</th><th>parent-of-origin</th><th>haplotype-of-origin</th><th>denovo_allele_sequence</th><th>precursor_allele_length_in_parent</th><th>precursor_sequence_in_parent</th><th>struc</th><th>allele_size_difference</th></tr><tr><td>str</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>str</td><td>str</td><td>u32</td></tr></thead><tbody><tr><td>&quot;chr10&quot;</td><td>101793189</td><td>101793216</td><td>&quot;mom:64&quot;</td><td>&quot;UNK&quot;</td><td>&quot;AAAAACAAAAACAAAAACAAAAACAAAAAC…</td><td>-1.0</td><td>&quot;UNK&quot;</td><td>&quot;(AAAAAC)n&quot;</td><td>36</td></tr><tr><td>&quot;chr10&quot;</td><td>30432978</td><td>30433009</td><td>&quot;dad:279&quot;</td><td>&quot;B&quot;</td><td>&quot;AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG…</td><td>247.0</td><td>&quot;AGAAGAGAGAGAGAGAGAGAGAGGAGAGGA…</td><td>&quot;(GA)n&quot;</td><td>4</td></tr><tr><td>&quot;chr10&quot;</td><td>30433093</td><td>30433105</td><td>&quot;unknown:0&quot;</td><td>&quot;UNK&quot;</td><td>&quot;GAGAGAGAGAGAGAGAGAGAGAGAGAGAGA…</td><td>-1.0</td><td>&quot;UNK&quot;</td><td>&quot;(AGGGGG)n&quot;</td><td>233</td></tr><tr><td>&quot;chr10&quot;</td><td>44662653</td><td>44663103</td><td>&quot;dad:445&quot;</td><td>&quot;B&quot;</td><td>&quot;AGGGGGCTGTGTCGTGGAGGGTGAGGATGG…</td><td>6050.0</td><td>&quot;AGGGGGCTGTGTCGTGGAGGGTGAGGATGG…</td><td>&quot;(GGGGTCCTAGGCGGCTGTGTCATGGAGGG…</td><td>150</td></tr><tr><td>&quot;chr10&quot;</td><td>66515265</td><td>66515370</td><td>&quot;mom:287&quot;</td><td>&quot;B&quot;</td><td>&quot;TATCTATCTATCTATCTATCTATCTATCTA…</td><td>129.0</td><td>&quot;TATCTATCTATCTATCTATCTATCTATCTA…</td><td>&quot;(TA)n(TC)n(TA)n&quot;</td><td>4294967284</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chr9&quot;</td><td>70469020</td><td>70469154</td><td>&quot;dad:144&quot;</td><td>&quot;B&quot;</td><td>&quot;TTTTCTTTTCTTCTCTTCTCTAATGACTTC…</td><td>134.0</td><td>&quot;TTTTCTTTTCTTCTCTTCTCTAATGACTTC…</td><td>&quot;(TTTC)n(TC)n(TTTC)n&quot;</td><td>109</td></tr><tr><td>&quot;chr9&quot;</td><td>96272373</td><td>96272473</td><td>&quot;dad:289&quot;</td><td>&quot;B&quot;</td><td>&quot;TCTCTCTCTCTCTCTCTCTCTCTCTCTCTC…</td><td>120.0</td><td>&quot;TCTCTCTCTCTCTCTCTCTCTCTCTCTCTC…</td><td>&quot;(TC)n(TA)n&quot;</td><td>4294967290</td></tr><tr><td>&quot;chrX&quot;</td><td>130516288</td><td>130516340</td><td>&quot;dad:123&quot;</td><td>&quot;UNK&quot;</td><td>&quot;AAATAAATAAATAAATAAATAAATAAATAA…</td><td>-1.0</td><td>&quot;UNK&quot;</td><td>&quot;(AAAT)n(GT)n&quot;</td><td>53</td></tr><tr><td>&quot;chrX&quot;</td><td>50967367</td><td>50967415</td><td>&quot;dad:52&quot;</td><td>&quot;UNK&quot;</td><td>&quot;ATATATATATATATATATATATATATATAT…</td><td>-1.0</td><td>&quot;UNK&quot;</td><td>&quot;(AT)n&quot;</td><td>53</td></tr><tr><td>&quot;chrX&quot;</td><td>68903896</td><td>68904208</td><td>&quot;dad:81&quot;</td><td>&quot;UNK&quot;</td><td>&quot;TTCTTTTCTTTTCTTTTCTTTTCTTTTCTT…</td><td>-1.0</td><td>&quot;UNK&quot;</td><td>&quot;(TTTC)n(T)n&quot;</td><td>271</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (91, 10)\n",
       "┌───────┬───────────┬───────────┬────────────┬───┬────────────┬────────────┬───────────┬───────────┐\n",
       "│ chrom ┆ start     ┆ end       ┆ parent-of- ┆ … ┆ precursor_ ┆ precursor_ ┆ struc     ┆ allele_si │\n",
       "│ ---   ┆ ---       ┆ ---       ┆ origin     ┆   ┆ allele_len ┆ sequence_i ┆ ---       ┆ ze_differ │\n",
       "│ str   ┆ i64       ┆ i64       ┆ ---        ┆   ┆ gth_in_par ┆ n_parent   ┆ str       ┆ ence      │\n",
       "│       ┆           ┆           ┆ str        ┆   ┆ …          ┆ ---        ┆           ┆ ---       │\n",
       "│       ┆           ┆           ┆            ┆   ┆ ---        ┆ str        ┆           ┆ u32       │\n",
       "│       ┆           ┆           ┆            ┆   ┆ f64        ┆            ┆           ┆           │\n",
       "╞═══════╪═══════════╪═══════════╪════════════╪═══╪════════════╪════════════╪═══════════╪═══════════╡\n",
       "│ chr10 ┆ 101793189 ┆ 101793216 ┆ mom:64     ┆ … ┆ -1.0       ┆ UNK        ┆ (AAAAAC)n ┆ 36        │\n",
       "│ chr10 ┆ 30432978  ┆ 30433009  ┆ dad:279    ┆ … ┆ 247.0      ┆ AGAAGAGAGA ┆ (GA)n     ┆ 4         │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ GAGAGAGAGA ┆           ┆           │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ GAGGAGAGGA ┆           ┆           │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ …          ┆           ┆           │\n",
       "│ chr10 ┆ 30433093  ┆ 30433105  ┆ unknown:0  ┆ … ┆ -1.0       ┆ UNK        ┆ (AGGGGG)n ┆ 233       │\n",
       "│ chr10 ┆ 44662653  ┆ 44663103  ┆ dad:445    ┆ … ┆ 6050.0     ┆ AGGGGGCTGT ┆ (GGGGTCCT ┆ 150       │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ GTCGTGGAGG ┆ AGGCGGCTG ┆           │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ GTGAGGATGG ┆ TGTCATGGA ┆           │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ …          ┆ GGG…      ┆           │\n",
       "│ chr10 ┆ 66515265  ┆ 66515370  ┆ mom:287    ┆ … ┆ 129.0      ┆ TATCTATCTA ┆ (TA)n(TC) ┆ 429496728 │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ TCTATCTATC ┆ n(TA)n    ┆ 4         │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ TATCTATCTA ┆           ┆           │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ …          ┆           ┆           │\n",
       "│ …     ┆ …         ┆ …         ┆ …          ┆ … ┆ …          ┆ …          ┆ …         ┆ …         │\n",
       "│ chr9  ┆ 70469020  ┆ 70469154  ┆ dad:144    ┆ … ┆ 134.0      ┆ TTTTCTTTTC ┆ (TTTC)n(T ┆ 109       │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ TTCTCTTCTC ┆ C)n(TTTC) ┆           │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ TAATGACTTC ┆ n         ┆           │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ …          ┆           ┆           │\n",
       "│ chr9  ┆ 96272373  ┆ 96272473  ┆ dad:289    ┆ … ┆ 120.0      ┆ TCTCTCTCTC ┆ (TC)n(TA) ┆ 429496729 │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ TCTCTCTCTC ┆ n         ┆ 0         │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ TCTCTCTCTC ┆           ┆           │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆ …          ┆           ┆           │\n",
       "│ chrX  ┆ 130516288 ┆ 130516340 ┆ dad:123    ┆ … ┆ -1.0       ┆ UNK        ┆ (AAAT)n(G ┆ 53        │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆            ┆ T)n       ┆           │\n",
       "│ chrX  ┆ 50967367  ┆ 50967415  ┆ dad:52     ┆ … ┆ -1.0       ┆ UNK        ┆ (AT)n     ┆ 53        │\n",
       "│ chrX  ┆ 68903896  ┆ 68904208  ┆ dad:81     ┆ … ┆ -1.0       ┆ UNK        ┆ (TTTC)n(T ┆ 271       │\n",
       "│       ┆           ┆           ┆            ┆   ┆            ┆            ┆ )n        ┆           │\n",
       "└───────┴───────────┴───────────┴────────────┴───┴────────────┴────────────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import polars as pl\n",
    "\n",
    "pl.Config.set_tbl_rows(10)\n",
    "\n",
    "# Tom Sasani provided TR DNMs for the child 2211: \n",
    "# https://quinlangroup.slack.com/archives/D9LFRMXV3/p1753988258836809\n",
    "CHILD_TR_DNMS = pl.read_csv(\"/scratch/ucgd/lustre-labs/quinlan/u1006375/CEPH-K1463-TandemRepeats/tr_validation/seq/2211.GRCh38.allele_sequences.tsv\", separator=\"\\t\")\n",
    "\n",
    "(\n",
    "    CHILD_TR_DNMS\n",
    "    .select(['#chrom', 'start', 'end', 'phase_consensus', 'haplotype_in_parent_consensus', 'denovo_allele_sequence', 'precursor_allele_length_in_parent', 'precursor_sequence_in_parent', 'struc'])\n",
    "    .with_columns([\n",
    "        (pl.col('denovo_allele_sequence').str.len_chars() - pl.col('precursor_sequence_in_parent').str.len_chars()).alias('allele_size_difference')\n",
    "    ])\n",
    "    .rename({\n",
    "        '#chrom': 'chrom',\n",
    "        'phase_consensus': 'parent-of-origin', \n",
    "        'haplotype_in_parent_consensus': 'haplotype-of-origin'\n",
    "    })\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1f3ed93",
   "metadata": {},
   "source": [
    "## Founder-phased DNA methylation at CpG sites common to Father, Mother, and Child in the Palladium trio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5bc127e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/scratch/ucgd/lustre-labs/quinlan/data-shared/dna-methylation/CEPH1463.GRCh38.hifi.model.founder-phased')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PB_CPG_TOOL_MODE = 'model' # mode of aligned_bam_to_cpg_scores\n",
    "# output dir of src/phase_meth_to_founder_haps.py\n",
    "OUTPUT_DIR = Path(f\"/scratch/ucgd/lustre-labs/quinlan/data-shared/dna-methylation/CEPH1463.GRCh38.hifi.{PB_CPG_TOOL_MODE}.founder-phased\")\n",
    "OUTPUT_DIR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2f13acd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (23_734_224, 7)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>start</th><th>end</th><th>methylation_level_pat</th><th>methylation_level_mat</th><th>founder_haplotype_pat</th><th>founder_haplotype_mat</th></tr><tr><td>str</td><td>i64</td><td>i64</td><td>f64</td><td>f64</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;chr1&quot;</td><td>689237</td><td>689238</td><td>0.228</td><td>0.927</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>689434</td><td>689435</td><td>0.075</td><td>0.577</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>690032</td><td>690033</td><td>0.889</td><td>0.905</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>690091</td><td>690092</td><td>0.87</td><td>0.76</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>690430</td><td>690431</td><td>0.761</td><td>0.045</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chr9&quot;</td><td>138216387</td><td>138216388</td><td>0.641</td><td>0.656</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td></tr><tr><td>&quot;chr9&quot;</td><td>138216400</td><td>138216401</td><td>0.89</td><td>0.927</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td></tr><tr><td>&quot;chr9&quot;</td><td>138216450</td><td>138216451</td><td>0.775</td><td>0.67</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td></tr><tr><td>&quot;chr9&quot;</td><td>138216484</td><td>138216485</td><td>0.658</td><td>0.547</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td></tr><tr><td>&quot;chr9&quot;</td><td>138216526</td><td>138216527</td><td>0.676</td><td>0.49</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (23_734_224, 7)\n",
       "┌───────┬───────────┬───────────┬────────────────┬────────────────┬────────────────┬───────────────┐\n",
       "│ chrom ┆ start     ┆ end       ┆ methylation_le ┆ methylation_le ┆ founder_haplot ┆ founder_haplo │\n",
       "│ ---   ┆ ---       ┆ ---       ┆ vel_pat        ┆ vel_mat        ┆ ype_pat        ┆ type_mat      │\n",
       "│ str   ┆ i64       ┆ i64       ┆ ---            ┆ ---            ┆ ---            ┆ ---           │\n",
       "│       ┆           ┆           ┆ f64            ┆ f64            ┆ str            ┆ str           │\n",
       "╞═══════╪═══════════╪═══════════╪════════════════╪════════════════╪════════════════╪═══════════════╡\n",
       "│ chr1  ┆ 689237    ┆ 689238    ┆ 0.228          ┆ 0.927          ┆ E              ┆ G             │\n",
       "│ chr1  ┆ 689434    ┆ 689435    ┆ 0.075          ┆ 0.577          ┆ E              ┆ G             │\n",
       "│ chr1  ┆ 690032    ┆ 690033    ┆ 0.889          ┆ 0.905          ┆ E              ┆ G             │\n",
       "│ chr1  ┆ 690091    ┆ 690092    ┆ 0.87           ┆ 0.76           ┆ E              ┆ G             │\n",
       "│ chr1  ┆ 690430    ┆ 690431    ┆ 0.761          ┆ 0.045          ┆ E              ┆ G             │\n",
       "│ …     ┆ …         ┆ …         ┆ …              ┆ …              ┆ …              ┆ …             │\n",
       "│ chr9  ┆ 138216387 ┆ 138216388 ┆ 0.641          ┆ 0.656          ┆ E              ┆ G             │\n",
       "│ chr9  ┆ 138216400 ┆ 138216401 ┆ 0.89           ┆ 0.927          ┆ E              ┆ G             │\n",
       "│ chr9  ┆ 138216450 ┆ 138216451 ┆ 0.775          ┆ 0.67           ┆ E              ┆ G             │\n",
       "│ chr9  ┆ 138216484 ┆ 138216485 ┆ 0.658          ┆ 0.547          ┆ E              ┆ G             │\n",
       "│ chr9  ┆ 138216526 ┆ 138216527 ┆ 0.676          ┆ 0.49           ┆ E              ┆ G             │\n",
       "└───────┴───────────┴───────────┴────────────────┴────────────────┴────────────────┴───────────────┘"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from read_data import read_data\n",
    "\n",
    "def get_methylation_sample(uid): \n",
    "    return (\n",
    "        read_data(OUTPUT_DIR, f\"{uid}.dna-methylation.founder-phased\")\n",
    "        .select(['chrom', 'start', 'end', 'methylation_level_pat', 'methylation_level_mat', 'founder_haplotype_pat', 'founder_haplotype_mat'])\n",
    "    )\n",
    "\n",
    "get_methylation_sample(FATHER_UID).filter(pl.col(\"founder_haplotype_pat\").is_not_null())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b5965522",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (15_749_868, 15)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>start</th><th>end</th><th>father_methylation_level_pat</th><th>father_methylation_level_mat</th><th>father_founder_haplotype_pat</th><th>father_founder_haplotype_mat</th><th>mother_methylation_level_pat</th><th>mother_methylation_level_mat</th><th>mother_founder_haplotype_pat</th><th>mother_founder_haplotype_mat</th><th>child_methylation_level_pat</th><th>child_methylation_level_mat</th><th>child_founder_haplotype_pat</th><th>child_founder_haplotype_mat</th></tr><tr><td>str</td><td>i64</td><td>i64</td><td>f64</td><td>f64</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;chr1&quot;</td><td>689434</td><td>689435</td><td>0.075</td><td>0.577</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td><td>0.764</td><td>0.695</td><td>&quot;I&quot;</td><td>&quot;K&quot;</td><td>0.586</td><td>0.097</td><td>&quot;G&quot;</td><td>&quot;K&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>690032</td><td>690033</td><td>0.889</td><td>0.905</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td><td>0.883</td><td>0.804</td><td>&quot;I&quot;</td><td>&quot;K&quot;</td><td>0.858</td><td>0.957</td><td>&quot;G&quot;</td><td>&quot;K&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>690091</td><td>690092</td><td>0.87</td><td>0.76</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td><td>0.796</td><td>0.866</td><td>&quot;I&quot;</td><td>&quot;K&quot;</td><td>0.902</td><td>0.924</td><td>&quot;G&quot;</td><td>&quot;K&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>690621</td><td>690622</td><td>0.785</td><td>0.544</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td><td>0.515</td><td>0.494</td><td>&quot;I&quot;</td><td>&quot;K&quot;</td><td>0.588</td><td>0.857</td><td>&quot;G&quot;</td><td>&quot;K&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>690673</td><td>690674</td><td>0.94</td><td>0.583</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td><td>0.75</td><td>0.904</td><td>&quot;I&quot;</td><td>&quot;K&quot;</td><td>0.634</td><td>0.928</td><td>&quot;G&quot;</td><td>&quot;K&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chr9&quot;</td><td>138216387</td><td>138216388</td><td>0.641</td><td>0.656</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td><td>0.646</td><td>0.918</td><td>&quot;I&quot;</td><td>&quot;K&quot;</td><td>0.889</td><td>0.659</td><td>&quot;E&quot;</td><td>&quot;K&quot;</td></tr><tr><td>&quot;chr9&quot;</td><td>138216400</td><td>138216401</td><td>0.89</td><td>0.927</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td><td>0.805</td><td>0.928</td><td>&quot;I&quot;</td><td>&quot;K&quot;</td><td>0.748</td><td>0.848</td><td>&quot;E&quot;</td><td>&quot;K&quot;</td></tr><tr><td>&quot;chr9&quot;</td><td>138216450</td><td>138216451</td><td>0.775</td><td>0.67</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td><td>0.845</td><td>0.546</td><td>&quot;I&quot;</td><td>&quot;K&quot;</td><td>0.903</td><td>0.399</td><td>&quot;E&quot;</td><td>&quot;K&quot;</td></tr><tr><td>&quot;chr9&quot;</td><td>138216484</td><td>138216485</td><td>0.658</td><td>0.547</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td><td>0.652</td><td>0.774</td><td>&quot;I&quot;</td><td>&quot;K&quot;</td><td>0.91</td><td>0.429</td><td>&quot;E&quot;</td><td>&quot;K&quot;</td></tr><tr><td>&quot;chr9&quot;</td><td>138216526</td><td>138216527</td><td>0.676</td><td>0.49</td><td>&quot;E&quot;</td><td>&quot;G&quot;</td><td>0.858</td><td>0.668</td><td>&quot;I&quot;</td><td>&quot;K&quot;</td><td>0.077</td><td>0.461</td><td>&quot;E&quot;</td><td>&quot;K&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (15_749_868, 15)\n",
       "┌───────┬───────────┬───────────┬────────────┬───┬────────────┬────────────┬───────────┬───────────┐\n",
       "│ chrom ┆ start     ┆ end       ┆ father_met ┆ … ┆ child_meth ┆ child_meth ┆ child_fou ┆ child_fou │\n",
       "│ ---   ┆ ---       ┆ ---       ┆ hylation_l ┆   ┆ ylation_le ┆ ylation_le ┆ nder_hapl ┆ nder_hapl │\n",
       "│ str   ┆ i64       ┆ i64       ┆ evel_pat   ┆   ┆ vel_pat    ┆ vel_mat    ┆ otype_pat ┆ otype_mat │\n",
       "│       ┆           ┆           ┆ ---        ┆   ┆ ---        ┆ ---        ┆ ---       ┆ ---       │\n",
       "│       ┆           ┆           ┆ f64        ┆   ┆ f64        ┆ f64        ┆ str       ┆ str       │\n",
       "╞═══════╪═══════════╪═══════════╪════════════╪═══╪════════════╪════════════╪═══════════╪═══════════╡\n",
       "│ chr1  ┆ 689434    ┆ 689435    ┆ 0.075      ┆ … ┆ 0.586      ┆ 0.097      ┆ G         ┆ K         │\n",
       "│ chr1  ┆ 690032    ┆ 690033    ┆ 0.889      ┆ … ┆ 0.858      ┆ 0.957      ┆ G         ┆ K         │\n",
       "│ chr1  ┆ 690091    ┆ 690092    ┆ 0.87       ┆ … ┆ 0.902      ┆ 0.924      ┆ G         ┆ K         │\n",
       "│ chr1  ┆ 690621    ┆ 690622    ┆ 0.785      ┆ … ┆ 0.588      ┆ 0.857      ┆ G         ┆ K         │\n",
       "│ chr1  ┆ 690673    ┆ 690674    ┆ 0.94       ┆ … ┆ 0.634      ┆ 0.928      ┆ G         ┆ K         │\n",
       "│ …     ┆ …         ┆ …         ┆ …          ┆ … ┆ …          ┆ …          ┆ …         ┆ …         │\n",
       "│ chr9  ┆ 138216387 ┆ 138216388 ┆ 0.641      ┆ … ┆ 0.889      ┆ 0.659      ┆ E         ┆ K         │\n",
       "│ chr9  ┆ 138216400 ┆ 138216401 ┆ 0.89       ┆ … ┆ 0.748      ┆ 0.848      ┆ E         ┆ K         │\n",
       "│ chr9  ┆ 138216450 ┆ 138216451 ┆ 0.775      ┆ … ┆ 0.903      ┆ 0.399      ┆ E         ┆ K         │\n",
       "│ chr9  ┆ 138216484 ┆ 138216485 ┆ 0.658      ┆ … ┆ 0.91       ┆ 0.429      ┆ E         ┆ K         │\n",
       "│ chr9  ┆ 138216526 ┆ 138216527 ┆ 0.676      ┆ … ┆ 0.077      ┆ 0.461      ┆ E         ┆ K         │\n",
       "└───────┴───────────┴───────────┴────────────┴───┴────────────┴────────────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def prefix_columns(df, prefix, join_keys):\n",
    "    # add prefixes to all columns except the join keys\n",
    "    return df.rename({col: f\"{prefix}_{col}\" for col in df.columns if col not in join_keys})\n",
    "\n",
    "def get_methylation_trio(father_uid, mother_uid, child_uid):\n",
    "    father_meth = get_methylation_sample(father_uid)\n",
    "    mother_meth = get_methylation_sample(mother_uid)\n",
    "    child_meth = get_methylation_sample(child_uid)\n",
    "    \n",
    "    father_prefixed = prefix_columns(father_meth, prefix=\"father\", join_keys=['chrom', 'start', 'end'])\n",
    "    mother_prefixed = prefix_columns(mother_meth, prefix=\"mother\", join_keys=['chrom', 'start', 'end'])\n",
    "    child_prefixed = prefix_columns(child_meth, prefix=\"child\", join_keys=['chrom', 'start', 'end'])\n",
    "\n",
    "    # Join on chrom, start, end\n",
    "    return (\n",
    "        father_prefixed\n",
    "        .join(mother_prefixed, on=['chrom', 'start', 'end'], how='inner')\n",
    "        .join(child_prefixed, on=['chrom', 'start', 'end'], how='inner')\n",
    "    )\n",
    "\n",
    "TRIO_METH = get_methylation_trio(FATHER_UID, MOTHER_UID, CHILD_UID)\n",
    "\n",
    "TRIO_METH.filter(pl.col(\"father_founder_haplotype_pat\").is_not_null())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b853200",
   "metadata": {},
   "source": [
    "## Computing a set of intervals in each of which all CpG sites share the same founder labels on each of the 6 haplotypes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9155102f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (3_088_273, 3)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>start</th><th>end</th></tr><tr><td>str</td><td>i64</td><td>i64</td></tr></thead><tbody><tr><td>&quot;chr1&quot;</td><td>0</td><td>1000</td></tr><tr><td>&quot;chr1&quot;</td><td>1000</td><td>2000</td></tr><tr><td>&quot;chr1&quot;</td><td>2000</td><td>3000</td></tr><tr><td>&quot;chr1&quot;</td><td>3000</td><td>4000</td></tr><tr><td>&quot;chr1&quot;</td><td>4000</td><td>5000</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chrM&quot;</td><td>11000</td><td>12000</td></tr><tr><td>&quot;chrM&quot;</td><td>12000</td><td>13000</td></tr><tr><td>&quot;chrM&quot;</td><td>13000</td><td>14000</td></tr><tr><td>&quot;chrM&quot;</td><td>14000</td><td>15000</td></tr><tr><td>&quot;chrM&quot;</td><td>15000</td><td>16000</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (3_088_273, 3)\n",
       "┌───────┬───────┬───────┐\n",
       "│ chrom ┆ start ┆ end   │\n",
       "│ ---   ┆ ---   ┆ ---   │\n",
       "│ str   ┆ i64   ┆ i64   │\n",
       "╞═══════╪═══════╪═══════╡\n",
       "│ chr1  ┆ 0     ┆ 1000  │\n",
       "│ chr1  ┆ 1000  ┆ 2000  │\n",
       "│ chr1  ┆ 2000  ┆ 3000  │\n",
       "│ chr1  ┆ 3000  ┆ 4000  │\n",
       "│ chr1  ┆ 4000  ┆ 5000  │\n",
       "│ …     ┆ …     ┆ …     │\n",
       "│ chrM  ┆ 11000 ┆ 12000 │\n",
       "│ chrM  ┆ 12000 ┆ 13000 │\n",
       "│ chrM  ┆ 13000 ┆ 14000 │\n",
       "│ chrM  ┆ 14000 ┆ 15000 │\n",
       "│ chrM  ┆ 15000 ┆ 16000 │\n",
       "└───────┴───────┴───────┘"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import bioframe as bf # https://bioframe.readthedocs.io/en/latest/index.html\n",
    "\n",
    "def get_tiles(reference_genome, size):\n",
    "    chromsizes = bf.fetch_chromsizes(db=reference_genome)\n",
    "    windows = []\n",
    "    for chrom, length in chromsizes.items():\n",
    "        for start in range(0, length, size): # type: ignore\n",
    "            end = start + size\n",
    "            if end > length:\n",
    "                continue\n",
    "            windows.append({\"chrom\": chrom, \"start\": start, \"end\": end})\n",
    "\n",
    "    return pl.DataFrame(windows)\n",
    "\n",
    "get_tiles(REFERENCE_GENOME, size=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b77cff33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (8_688, 3)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>start</th><th>end</th></tr><tr><td>str</td><td>i64</td><td>i64</td></tr></thead><tbody><tr><td>&quot;chr1&quot;</td><td>668466</td><td>1052726</td></tr><tr><td>&quot;chr1&quot;</td><td>1074443</td><td>1079281</td></tr><tr><td>&quot;chr1&quot;</td><td>1108637</td><td>1265603</td></tr><tr><td>&quot;chr1&quot;</td><td>1288373</td><td>1610923</td></tr><tr><td>&quot;chr1&quot;</td><td>1613420</td><td>2370170</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chr9&quot;</td><td>137873339</td><td>137943538</td></tr><tr><td>&quot;chr9&quot;</td><td>137976541</td><td>137977195</td></tr><tr><td>&quot;chr9&quot;</td><td>138002413</td><td>138087950</td></tr><tr><td>&quot;chr9&quot;</td><td>138111774</td><td>138174171</td></tr><tr><td>&quot;chr9&quot;</td><td>138174280</td><td>138216552</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (8_688, 3)\n",
       "┌───────┬───────────┬───────────┐\n",
       "│ chrom ┆ start     ┆ end       │\n",
       "│ ---   ┆ ---       ┆ ---       │\n",
       "│ str   ┆ i64       ┆ i64       │\n",
       "╞═══════╪═══════════╪═══════════╡\n",
       "│ chr1  ┆ 668466    ┆ 1052726   │\n",
       "│ chr1  ┆ 1074443   ┆ 1079281   │\n",
       "│ chr1  ┆ 1108637   ┆ 1265603   │\n",
       "│ chr1  ┆ 1288373   ┆ 1610923   │\n",
       "│ chr1  ┆ 1613420   ┆ 2370170   │\n",
       "│ …     ┆ …         ┆ …         │\n",
       "│ chr9  ┆ 137873339 ┆ 137943538 │\n",
       "│ chr9  ┆ 137976541 ┆ 137977195 │\n",
       "│ chr9  ┆ 138002413 ┆ 138087950 │\n",
       "│ chr9  ┆ 138111774 ┆ 138174171 │\n",
       "│ chr9  ┆ 138174280 ┆ 138216552 │\n",
       "└───────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_hap_map_blocks(uid):\n",
    "    return (\n",
    "        read_data(OUTPUT_DIR, f\"{uid}.hap-map-blocks\")\n",
    "        .select(['chrom', 'start', 'end'])\n",
    "    )\n",
    "\n",
    "get_hap_map_blocks(FATHER_UID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4c57141f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (1_915_796, 3)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>start</th><th>end</th></tr><tr><td>str</td><td>i64</td><td>i64</td></tr></thead><tbody><tr><td>&quot;chr17&quot;</td><td>119000</td><td>120000</td></tr><tr><td>&quot;chr17&quot;</td><td>120000</td><td>121000</td></tr><tr><td>&quot;chr17&quot;</td><td>121000</td><td>122000</td></tr><tr><td>&quot;chr17&quot;</td><td>122000</td><td>123000</td></tr><tr><td>&quot;chr17&quot;</td><td>123000</td><td>124000</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chr10&quot;</td><td>133635000</td><td>133636000</td></tr><tr><td>&quot;chr10&quot;</td><td>133636000</td><td>133637000</td></tr><tr><td>&quot;chr10&quot;</td><td>133637000</td><td>133638000</td></tr><tr><td>&quot;chr10&quot;</td><td>133638000</td><td>133639000</td></tr><tr><td>&quot;chr10&quot;</td><td>133639000</td><td>133640000</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (1_915_796, 3)\n",
       "┌───────┬───────────┬───────────┐\n",
       "│ chrom ┆ start     ┆ end       │\n",
       "│ ---   ┆ ---       ┆ ---       │\n",
       "│ str   ┆ i64       ┆ i64       │\n",
       "╞═══════╪═══════════╪═══════════╡\n",
       "│ chr17 ┆ 119000    ┆ 120000    │\n",
       "│ chr17 ┆ 120000    ┆ 121000    │\n",
       "│ chr17 ┆ 121000    ┆ 122000    │\n",
       "│ chr17 ┆ 122000    ┆ 123000    │\n",
       "│ chr17 ┆ 123000    ┆ 124000    │\n",
       "│ …     ┆ …         ┆ …         │\n",
       "│ chr10 ┆ 133635000 ┆ 133636000 │\n",
       "│ chr10 ┆ 133636000 ┆ 133637000 │\n",
       "│ chr10 ┆ 133637000 ┆ 133638000 │\n",
       "│ chr10 ┆ 133638000 ┆ 133639000 │\n",
       "│ chr10 ┆ 133639000 ┆ 133640000 │\n",
       "└───────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_intervals(reference_genome, size, father_uid, mother_uid, child_uid):\n",
    "    tiles = get_tiles(reference_genome, size).to_pandas()\n",
    "    father_blocks = get_hap_map_blocks(father_uid).to_pandas()\n",
    "    mother_blocks = get_hap_map_blocks(mother_uid).to_pandas()\n",
    "    child_blocks = get_hap_map_blocks(child_uid).to_pandas()\n",
    "\n",
    "    # Intersect tiles and father blocks using bioframe.overlap\n",
    "    tiles_father = (\n",
    "        bf.overlap(\n",
    "            tiles,\n",
    "            father_blocks,\n",
    "            how='inner',\n",
    "            return_overlap=True\n",
    "        )\n",
    "        [[\"chrom\", \"overlap_start\", \"overlap_end\"]]\n",
    "        .rename(columns={\"overlap_start\": \"start\", \"overlap_end\": \"end\"})\n",
    "    )\n",
    "\n",
    "    # Intersect tiles_father and mother blocks using bioframe.overlap\n",
    "    tiles_father_mother = (\n",
    "        bf.overlap(\n",
    "            tiles_father,\n",
    "            mother_blocks,\n",
    "            how='inner',\n",
    "            return_overlap=True\n",
    "        )\n",
    "        [[\"chrom\", \"overlap_start\", \"overlap_end\"]]\n",
    "        .rename(columns={\"overlap_start\": \"start\", \"overlap_end\": \"end\"})\n",
    "    )\n",
    "\n",
    "    # Intersect tiles_father_mother and child blocks using bioframe.overlap\n",
    "    tiles_father_mother_child = (\n",
    "        bf.overlap(\n",
    "            tiles_father_mother,\n",
    "            child_blocks,\n",
    "            how='inner',\n",
    "            return_overlap=True\n",
    "        )\n",
    "        [[\"chrom\", \"overlap_start\", \"overlap_end\"]]\n",
    "        .rename(columns={\"overlap_start\": \"start\", \"overlap_end\": \"end\"})\n",
    "    )\n",
    "\n",
    "    return (\n",
    "        pl\n",
    "        .from_pandas(tiles_father_mother_child)\n",
    "        .filter(pl.col(\"end\") - pl.col(\"start\") == size)\n",
    "    ) \n",
    "\n",
    "INTERVALS = get_intervals(REFERENCE_GENOME, size=1000, father_uid=FATHER_UID, mother_uid=MOTHER_UID, child_uid=CHILD_UID)\n",
    "INTERVALS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6e85f09",
   "metadata": {},
   "source": [
    "## XXX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21917b41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>father_methylation_level_pat</th>\n",
       "      <th>father_methylation_level_mat</th>\n",
       "      <th>father_founder_haplotype_pat</th>\n",
       "      <th>father_founder_haplotype_mat</th>\n",
       "      <th>mother_methylation_level_pat</th>\n",
       "      <th>mother_methylation_level_mat</th>\n",
       "      <th>mother_founder_haplotype_pat</th>\n",
       "      <th>mother_founder_haplotype_mat</th>\n",
       "      <th>child_methylation_level_pat</th>\n",
       "      <th>child_methylation_level_mat</th>\n",
       "      <th>child_founder_haplotype_pat</th>\n",
       "      <th>child_founder_haplotype_mat</th>\n",
       "      <th>chrom_</th>\n",
       "      <th>start_</th>\n",
       "      <th>end_</th>\n",
       "      <th>overlap_start</th>\n",
       "      <th>overlap_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr17</td>\n",
       "      <td>130000</td>\n",
       "      <td>130001</td>\n",
       "      <td>0.938</td>\n",
       "      <td>0.954</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>0.823</td>\n",
       "      <td>0.946</td>\n",
       "      <td>I</td>\n",
       "      <td>K</td>\n",
       "      <td>0.873</td>\n",
       "      <td>0.885</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>chr17</td>\n",
       "      <td>130000</td>\n",
       "      <td>131000</td>\n",
       "      <td>130000</td>\n",
       "      <td>130001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>139000</td>\n",
       "      <td>139001</td>\n",
       "      <td>0.954</td>\n",
       "      <td>0.939</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>0.949</td>\n",
       "      <td>0.941</td>\n",
       "      <td>I</td>\n",
       "      <td>K</td>\n",
       "      <td>0.928</td>\n",
       "      <td>0.893</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>chr17</td>\n",
       "      <td>139000</td>\n",
       "      <td>140000</td>\n",
       "      <td>139000</td>\n",
       "      <td>139001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>142000</td>\n",
       "      <td>142001</td>\n",
       "      <td>0.917</td>\n",
       "      <td>0.969</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>0.932</td>\n",
       "      <td>0.952</td>\n",
       "      <td>I</td>\n",
       "      <td>K</td>\n",
       "      <td>0.915</td>\n",
       "      <td>0.910</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>chr17</td>\n",
       "      <td>142000</td>\n",
       "      <td>143000</td>\n",
       "      <td>142000</td>\n",
       "      <td>142001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17</td>\n",
       "      <td>797000</td>\n",
       "      <td>797001</td>\n",
       "      <td>0.948</td>\n",
       "      <td>0.927</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>0.907</td>\n",
       "      <td>0.926</td>\n",
       "      <td>I</td>\n",
       "      <td>K</td>\n",
       "      <td>0.957</td>\n",
       "      <td>0.955</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>chr17</td>\n",
       "      <td>797000</td>\n",
       "      <td>798000</td>\n",
       "      <td>797000</td>\n",
       "      <td>797001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17</td>\n",
       "      <td>978000</td>\n",
       "      <td>978001</td>\n",
       "      <td>0.261</td>\n",
       "      <td>0.584</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>0.431</td>\n",
       "      <td>0.580</td>\n",
       "      <td>I</td>\n",
       "      <td>K</td>\n",
       "      <td>0.365</td>\n",
       "      <td>0.757</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>chr17</td>\n",
       "      <td>978000</td>\n",
       "      <td>979000</td>\n",
       "      <td>978000</td>\n",
       "      <td>978001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13932432</th>\n",
       "      <td>chr10</td>\n",
       "      <td>133639311</td>\n",
       "      <td>133639312</td>\n",
       "      <td>0.878</td>\n",
       "      <td>0.942</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>0.937</td>\n",
       "      <td>0.945</td>\n",
       "      <td>I</td>\n",
       "      <td>K</td>\n",
       "      <td>0.934</td>\n",
       "      <td>0.889</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>chr10</td>\n",
       "      <td>133639000</td>\n",
       "      <td>133640000</td>\n",
       "      <td>133639311</td>\n",
       "      <td>133639312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13932433</th>\n",
       "      <td>chr10</td>\n",
       "      <td>133639660</td>\n",
       "      <td>133639661</td>\n",
       "      <td>0.937</td>\n",
       "      <td>0.931</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>0.938</td>\n",
       "      <td>0.934</td>\n",
       "      <td>I</td>\n",
       "      <td>K</td>\n",
       "      <td>0.935</td>\n",
       "      <td>0.946</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>chr10</td>\n",
       "      <td>133639000</td>\n",
       "      <td>133640000</td>\n",
       "      <td>133639660</td>\n",
       "      <td>133639661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13932434</th>\n",
       "      <td>chr10</td>\n",
       "      <td>133639705</td>\n",
       "      <td>133639706</td>\n",
       "      <td>0.953</td>\n",
       "      <td>0.946</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>0.950</td>\n",
       "      <td>0.938</td>\n",
       "      <td>I</td>\n",
       "      <td>K</td>\n",
       "      <td>0.957</td>\n",
       "      <td>0.959</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>chr10</td>\n",
       "      <td>133639000</td>\n",
       "      <td>133640000</td>\n",
       "      <td>133639705</td>\n",
       "      <td>133639706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13932435</th>\n",
       "      <td>chr10</td>\n",
       "      <td>133639968</td>\n",
       "      <td>133639969</td>\n",
       "      <td>0.941</td>\n",
       "      <td>0.951</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>0.958</td>\n",
       "      <td>0.890</td>\n",
       "      <td>I</td>\n",
       "      <td>K</td>\n",
       "      <td>0.942</td>\n",
       "      <td>0.535</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>chr10</td>\n",
       "      <td>133639000</td>\n",
       "      <td>133640000</td>\n",
       "      <td>133639968</td>\n",
       "      <td>133639969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13932436</th>\n",
       "      <td>chr10</td>\n",
       "      <td>133639983</td>\n",
       "      <td>133639984</td>\n",
       "      <td>0.956</td>\n",
       "      <td>0.959</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>0.959</td>\n",
       "      <td>0.958</td>\n",
       "      <td>I</td>\n",
       "      <td>K</td>\n",
       "      <td>0.958</td>\n",
       "      <td>0.939</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>chr10</td>\n",
       "      <td>133639000</td>\n",
       "      <td>133640000</td>\n",
       "      <td>133639983</td>\n",
       "      <td>133639984</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13932437 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom      start        end  father_methylation_level_pat  \\\n",
       "0         chr17     130000     130001                         0.938   \n",
       "1         chr17     139000     139001                         0.954   \n",
       "2         chr17     142000     142001                         0.917   \n",
       "3         chr17     797000     797001                         0.948   \n",
       "4         chr17     978000     978001                         0.261   \n",
       "...         ...        ...        ...                           ...   \n",
       "13932432  chr10  133639311  133639312                         0.878   \n",
       "13932433  chr10  133639660  133639661                         0.937   \n",
       "13932434  chr10  133639705  133639706                         0.953   \n",
       "13932435  chr10  133639968  133639969                         0.941   \n",
       "13932436  chr10  133639983  133639984                         0.956   \n",
       "\n",
       "          father_methylation_level_mat father_founder_haplotype_pat  \\\n",
       "0                                0.954                            E   \n",
       "1                                0.939                            E   \n",
       "2                                0.969                            E   \n",
       "3                                0.927                            E   \n",
       "4                                0.584                            E   \n",
       "...                                ...                          ...   \n",
       "13932432                         0.942                            E   \n",
       "13932433                         0.931                            E   \n",
       "13932434                         0.946                            E   \n",
       "13932435                         0.951                            E   \n",
       "13932436                         0.959                            E   \n",
       "\n",
       "         father_founder_haplotype_mat  mother_methylation_level_pat  \\\n",
       "0                                   G                         0.823   \n",
       "1                                   G                         0.949   \n",
       "2                                   G                         0.932   \n",
       "3                                   G                         0.907   \n",
       "4                                   G                         0.431   \n",
       "...                               ...                           ...   \n",
       "13932432                            G                         0.937   \n",
       "13932433                            G                         0.938   \n",
       "13932434                            G                         0.950   \n",
       "13932435                            G                         0.958   \n",
       "13932436                            G                         0.959   \n",
       "\n",
       "          mother_methylation_level_mat mother_founder_haplotype_pat  \\\n",
       "0                                0.946                            I   \n",
       "1                                0.941                            I   \n",
       "2                                0.952                            I   \n",
       "3                                0.926                            I   \n",
       "4                                0.580                            I   \n",
       "...                                ...                          ...   \n",
       "13932432                         0.945                            I   \n",
       "13932433                         0.934                            I   \n",
       "13932434                         0.938                            I   \n",
       "13932435                         0.890                            I   \n",
       "13932436                         0.958                            I   \n",
       "\n",
       "         mother_founder_haplotype_mat  child_methylation_level_pat  \\\n",
       "0                                   K                        0.873   \n",
       "1                                   K                        0.928   \n",
       "2                                   K                        0.915   \n",
       "3                                   K                        0.957   \n",
       "4                                   K                        0.365   \n",
       "...                               ...                          ...   \n",
       "13932432                            K                        0.934   \n",
       "13932433                            K                        0.935   \n",
       "13932434                            K                        0.957   \n",
       "13932435                            K                        0.942   \n",
       "13932436                            K                        0.958   \n",
       "\n",
       "          child_methylation_level_mat child_founder_haplotype_pat  \\\n",
       "0                               0.885                           G   \n",
       "1                               0.893                           G   \n",
       "2                               0.910                           G   \n",
       "3                               0.955                           G   \n",
       "4                               0.757                           G   \n",
       "...                               ...                         ...   \n",
       "13932432                        0.889                           G   \n",
       "13932433                        0.946                           G   \n",
       "13932434                        0.959                           G   \n",
       "13932435                        0.535                           G   \n",
       "13932436                        0.939                           G   \n",
       "\n",
       "         child_founder_haplotype_mat chrom_     start_       end_  \\\n",
       "0                                  K  chr17     130000     131000   \n",
       "1                                  K  chr17     139000     140000   \n",
       "2                                  K  chr17     142000     143000   \n",
       "3                                  K  chr17     797000     798000   \n",
       "4                                  K  chr17     978000     979000   \n",
       "...                              ...    ...        ...        ...   \n",
       "13932432                           K  chr10  133639000  133640000   \n",
       "13932433                           K  chr10  133639000  133640000   \n",
       "13932434                           K  chr10  133639000  133640000   \n",
       "13932435                           K  chr10  133639000  133640000   \n",
       "13932436                           K  chr10  133639000  133640000   \n",
       "\n",
       "          overlap_start  overlap_end  \n",
       "0                130000       130001  \n",
       "1                139000       139001  \n",
       "2                142000       142001  \n",
       "3                797000       797001  \n",
       "4                978000       978001  \n",
       "...                 ...          ...  \n",
       "13932432      133639311    133639312  \n",
       "13932433      133639660    133639661  \n",
       "13932434      133639705    133639706  \n",
       "13932435      133639968    133639969  \n",
       "13932436      133639983    133639984  \n",
       "\n",
       "[13932437 rows x 20 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def f(trio_meth, intervals): \n",
    "    return bf.overlap(\n",
    "        trio_meth.to_pandas(),\n",
    "        intervals.to_pandas(),\n",
    "        how='inner',\n",
    "        return_overlap=False\n",
    "    )\n",
    "\n",
    "f(TRIO_METH, INTERVALS)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2fbcbb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO \n",
    "# 2. group by those intervals AND founder haplotypes (which are now, by construction, all the same at all CpG sites in each interval -- check this), and aggregate the methylation levels \n",
    "# 3. for each record, use founder haplotypes to pair each of the child methylation levels to a parental methylation level, and compute the diff, as follows:  \n",
    "#     i. determine which of father_founder_haplotype_pat, father_founder_haplotype_mat, mother_founder_haplotype_pat or mother_founder_haplotype_mat matches child_founder_haplotype_pat\n",
    "#     ii. e.g., if it is father_founder_haplotype_pat, then use father_methylation_level_pat and child_founder_haplotype_pat to compute the diff of methylation levels, etc.  \n",
    "#     iii. repeats steps i and ii for child_founder_haplotype_mat\n",
    "# 4. order blocks by diff of methylation levels for child_founder_haplotype_pat and child_founder_haplotype_mat, \n",
    "#    and inspect the top-ranked blocks in IGV \n",
    "#    (another way of saying look for outliers in diff of methylation levels, i.e., windows whose diffs lie in the tail of the distribution)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3ecccad",
   "metadata": {},
   "source": [
    "## Searching for epimutations"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
